### ISSUE TITLE
Unexpected readLine() / printing behaviour

### ISSUE THREAD

**Issue by ghost** [ziglang/zig#1055]
```
pub fn main() !void {
    const std = @import("std");
    const warn = std.debug.warn;    
    const io = std.io;
    var input: [300]u8 = undefined;
    const line_len = try io.readLine(input[0..]);
    warn("{} end", input);   
}
```
I would expect this code to echo back the content that is entered followed by the string " end". But it also prints the contents of the empty locations in the `input` variable.
Output - 
```
12345
12345                                                                                                                                                                                                                                                                                                        end
```

This is real surprising behaviour that you need to do `warn("{} end", input[0..line_len]);` for it to work as expected.

IMO zig really needs a proper `string` type as this current behaviour is seriously going backwards even from C.

**Comment by BraedonWooding** [comment]
You sure it doesn't work...?  Using try it outline it printed what you would expect, you can view it [here](https://tio.run/##LY5BCoMwEEX3OcU0K4US1G5KQ6EH6A1EijZRBjSRmFha8expos5u3v9//vyw8/6t1WRhsgLu8MBh1MYmNKw05WTXPrVRQQyMCdm4jkXAIcxhQH3IqDkho2ugVTDUqJIUTrNGAQuJ9rk2gGp09gblJcsqdw05p4RsUUnBN89@sQ/g1ctYa803FDAja/EMNNkOlBljVbon4jcJXVaQStDzXpByIKv3eV4U5A8).  (Look under the 'debug' section since its a debug print)

**Comment by Hejsil** [comment]
Also, `stdin` can accept any sequence of bytes, even invalid UTF8 or ascii, so it would not be correct to return a "string", unless you want `readLine` to validate the input before giving you the result.

What might be useful though, would be for `readLine` to return `[]u8` instead of `usize`. Then your example becomes:
```
pub fn main() !void {
    const std = @import("std");
    const warn = std.debug.warn;    
    const io = std.io;
    var input: [300]u8 = undefined;
    const line = try io.readLine(input[0..]);
    warn("{} end", line);   
}
```

**Comment by tiehuis** [comment]
So the main difference here versus C is that we don't use null-terminated strings for input, so using the buffer directly isn't going to be a direct analogue like with `fgets`, where the fgets call is implicitly specifying the length by null terminating the input in the function. We don't null-terminate buffers, so we need to return the length independently to know how much was read instead.

Some other alternative things we can do:

 - Return a slice representing the read buffer. Then you could just do `const line = try io.readLine(input[0..]);` which may be a little more obvious.
 - Take a growable `ArrayList` (preferably a Utf8String) and pass that to the function. The length can be changed within it and it matches your expectation.

There a some tradeoffs in these approaches and they are useful in different scenarios. I think the main issue here is that the lack of stdlib documentation/usage for various stdlib functions means there is a barrier for new users.

https://github.com/ziglang/zig/issues/965

**Comment by NeeleshChandola** [comment]
> Of course if you give it extra spaces in input it'll print those out since that's what has been 'read in'.
  
And that is the point. The function is named readLine() so it should only read upto a '\n' and therefore the buffer should only contain upto that and therefore only that should be printed. Coming from other languages that is the default behaviour you would expect.

I also agree that it would be best for the function to return a []u8 directly.
const line: []u8 = try io.readLine();

**Comment by BraedonWooding** [comment]
... Why would the buffer only contain that though?  You set a buffer of 100 long then expect the buffer to change its length on you?  To me that just sounds less conventional and obscure.  If you want null terminated strings there are std lib functions to convert to them, and you could go that route.

Also your confusing what the signature would be it would be this;
`const line: []u8 = try io.readline(buf);` where you still have to give it a buffer, else how would it create the memory?  There is no 'standard malloc' in Zig so either you have to pass it an allocator or a buffer.

**Comment by andrewrk** [comment]
I agree that this is a footgun. Let's update the standard library so that more users do not run into this problem.

**Comment by Hejsil** [comment]
Fixed #1801
